### PR TITLE
expression: implement vectorized evaluation for 'builtinCastJSONAsDurationSig'

### PR DESCRIPTION
**Pull request by tsthght**
### What problem does this PR solve? 
Implement vectorized evaluation for builtinCastRealAsStringSig, for #12176

### What is changed and how it works?
```
BenchmarkVectorizedBuiltinCastFunc/builtinCastJSONAsDurationSig-VecBuiltinFunc-8         	    2000	   1137558 ns/op	  221290 B/op	    9216 allocs/op
BenchmarkVectorizedBuiltinCastFunc/builtinCastJSONAsDurationSig-NonVecBuiltinFunc-8      	    2000	   1125579 ns/op	  221292 B/op	    9216 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_cast_vec.go [hunk 1]
```diff
@@ -833,11 +833,44 @@ func (b *builtinCastStringAsStringSig) vecEvalString(input *chunk.Chunk, result
...
+			continue
+		}
+		s, err := buf.GetJSON(i).Unquote()
+		if err != nil {
+			return nil
+		}
+		dur, err = types.ParseDuration(ctx, s, int8(b.tp.Decimal))
+		if types.ErrTruncatedWrongVal.Equal(err) {
+			sc := b.ctx.GetSessionVars().StmtCtx
+			err = sc.HandleTruncate(err)
```
Comment (SunRunAway) [inline expression/builtin_cast_vec.go]:
```suggestion
			err = ctx.HandleTruncate(err)
```

File: expression/bench_test.go [hunk 2]
```diff
@@ -416,6 +416,18 @@ func (g *randDurInt) gen() interface{} {
 	return int64(rand.Intn(types.TimeMaxHour)*10000 + rand.Intn(60)*100 + rand.Intn(60))
 }
 
+type randJsonDurationWithLen struct {
+	len int32
+}
+
+func (g *randJsonDurationWithLen) gen() interface{} {
+	j := new(json.BinaryJSON)
+	if err := j.UnmarshalJSON([]byte(fmt.Sprintf(`{"key":%v}`, rand.Int31n(g.len)))); err != nil {
```
Comment (SunRunAway) [inline expression/bench_test.go]: Why is it the form like `{key: <duration>}` but not `<duration>`? Does `builtinCastJSONAsDurationSig` have able to extract `<duration>` from the `key`?

File: expression/builtin_cast_vec.go (hunk 1 above)
Comment (tsthght) [reply expression/builtin_cast_vec.go]: ok, i will rename the parameter name

File: expression/bench_test.go (hunk 2 above)
Comment (tsthght) [reply expression/bench_test.go]: i think builtinCastJSONAsDurationSig indeed have able to extract duration from the JSON。The non vectorized implement like: 

func (b *builtinCastJSONAsDurationSig) evalDuration(row chunk.Row) (res types.Duration, isNull bool, err error) {
	val, isNull, err := b.args[0].EvalJSON(b.ctx, row)
	if isNull || err != nil {
		return res, isNull, err
	}
	s, err := val.Unquote()
	if err != nil {
		return res, false, err
	}
	res, err = types.ParseDuration(b.ctx.GetSessionVars().StmtCtx, s, int8(b.tp.Decimal))
	if types.ErrTruncatedWrongVal.Equal(err) {
		sc := b.ctx.GetSessionVars().StmtCtx
		err = sc.HandleTruncate(err)
	}
	return
}

**Comment by tsthght** [comment]
> I wonder why the vectorized implementation is slower than the old row-based implementation.
> Could you please do some investigation about this?
> You can use golang pprof tool like `go test -v -benchmem -run=xxx -bench=xxx -memprofile=mem.prof -cpuprofile=cpu.prof -args xxx` to get some profiles and use the pprof tool `go tool pprof mem.prof/cpu.prof` to open and analyze it.

1. The cost times of vectorized implementation is  similar to old row-based implementation:

image

2. I think the mainly cost is "parse JSON" 。So even using the vectorized implementation, the costs reduction is not significant.

CPU:
image

MEM:
image

File: expression/bench_test.go (hunk 2 above)
Comment (SunRunAway) [reply expression/bench_test.go]: I don't think so.
After executing `val, isNull, err := b.args[0].EvalJSON(b.ctx, row)`, the `val` is `{"key": 11}` which is a JSON type.
After executing `val.Unquote()`, the `s` is `{"key": 11}` which is a string type.
Then types.ParseDuration parses `{"key": 11}` to a duration.

Comment (SunRunAway) [reply expression/bench_test.go]: I don't think so.
After executing `val, isNull, err := b.args[0].EvalJSON(b.ctx, row)`, the `val` is `{"key": 11}` which is a JSON type.
After executing `val.Unquote()`, the `s` is `{"key": 11}` which is a string type.
Then types.ParseDuration parses `{"key": 11}` to a duration.

File: expression/builtin_cast_vec.go [hunk 3]
```diff
@@ -997,11 +997,44 @@ func (b *builtinCastStringAsStringSig) vecEvalString(input *chunk.Chunk, result
...
+		if result.IsNull(i) {
+			continue
+		}
+		s, err := buf.GetJSON(i).Unquote()
+		if err != nil {
+			return nil
+		}
+		dur, err = types.ParseDuration(ctx, s, int8(b.tp.Decimal))
+		if types.ErrTruncatedWrongVal.Equal(err) {
+			ctx := b.ctx.GetSessionVars().StmtCtx
```
Comment (SunRunAway) [inline expression/builtin_cast_vec.go]: This line is redundant.

Comment (tsthght) [reply expression/builtin_cast_vec.go]: I had removed this line, thank you~

**Comment by SunRunAway** [comment]
Hi, @tsthght 
Please resolve [this comment](https://github.com/pingcap/tidb/pull/12628#discussion_r335844612).

File: expression/bench_test.go (hunk 2 above)
Comment (tsthght) [reply expression/bench_test.go]: Sorry, after too long to reply. I had modified it. PTAL @SunRunAway @qw4990

File: expression/bench_test.go [hunk 4]
```diff
@@ -472,6 +472,15 @@ func (g *randDurInt) gen() interface{} {
 	return int64(rand.Intn(types.TimeMaxHour)*10000 + rand.Intn(60)*100 + rand.Intn(60))
 }
 
+type randJSONDurationWithLen struct{}
```
Comment (SunRunAway) [inline expression/bench_test.go]: Because it's not a common generator, could we put it in builtin_cast_vec_test.go?

Comment (SunRunAway) [reply expression/bench_test.go]: And we may change `randJSONDurationWithLen` to `randJSONDuration`.

Comment (qw4990) [reply expression/bench_test.go]: @tsthght Friendly ping, please address this comment.

Comment (tsthght) [reply expression/bench_test.go]: I had modified. PTAL @SunRunAway @qw4990

**Pull request merged**